vsintegration/src/FSharp.Editor/FSharpBraceMatchingService.fs

```diff
@@ -0,0 +1,115 @@
+
+namespace Microsoft.VisualStudio.FSharp.Editor
+
+open System
+open System.Composition
+open System.Collections.Concurrent
+open System.Collections.Generic
+open System.Threading
+open System.Threading.Tasks
+open System.Linq
+
+open Microsoft.CodeAnalysis
+open Microsoft.CodeAnalysis.Classification
+open Microsoft.CodeAnalysis.Editor
+open Microsoft.CodeAnalysis.Editor.Implementation.BraceMatching
+open Microsoft.CodeAnalysis.Editor.Shared.Utilities
+open Microsoft.CodeAnalysis.Host.Mef
+open Microsoft.CodeAnalysis.Text
+
+open Microsoft.VisualStudio.FSharp.LanguageService
+open Microsoft.VisualStudio.Text
+open Microsoft.VisualStudio.Text.Tagging
+
+open Microsoft.FSharp.Compiler.Parser
+open Microsoft.FSharp.Compiler.SourceCodeServices
+
+[<ExportBraceMatcher(FSharpCommonConstants.FSharpLanguageName)>]
+type internal FSharpBraceMatchingService() =
+              
+    static member private SupportedBraceTypes = [
+        ('(', ')');
+        ('<', '>');
+        ('[', ']');
+        ('{', '}');
+    ]
+    
+    static member private IgnoredClassificationTypes = [
+        ClassificationTypeNames.Comment;
+        ClassificationTypeNames.StringLiteral;
+        ClassificationTypeNames.ExcludedCode;
+    ]
+
+    interface IBraceMatcher with
+        member this.FindBracesAsync(document: Document, position: int, cancellationToken: CancellationToken): Task<Nullable<BraceMatchingResult>> =
+            let computation = async {
+                let! sourceText = document.GetTextAsync(cancellationToken) |> Async.AwaitTask
+                return
+                    try
+                        FSharpBraceMatchingService.GetBraceMatchingResult(sourceText, document.Name, position, cancellationToken)
+                    with ex -> 
+                        Assert.Exception(ex)
+                        reraise()
+            }
+            Async.StartAsTask(computation, TaskCreationOptions.None, cancellationToken)
+  
+    // Helper function to proxy Roslyn types to tests
+    static member FindMatchingBrace(sourceText: SourceText, fileName: string, position: int, cancellationToken: CancellationToken) : Option<int> =
+        let braceMatchingResult = FSharpBraceMatchingService.GetBraceMatchingResult(sourceText, fileName, position, cancellationToken)
+        if braceMatchingResult.HasValue then
+            if braceMatchingResult.Value.LeftSpan.Start = position then
+                Some(braceMatchingResult.Value.RightSpan.Start)
+            else if braceMatchingResult.Value.RightSpan.Start = position then
+                Some(braceMatchingResult.Value.LeftSpan.Start)
+            else
+                None
+        else
+            None
+
+    static member private GetBraceMatchingResult(sourceText: SourceText, fileName: string, position: int, cancellationToken: CancellationToken) : Nullable<BraceMatchingResult> =
```

Comment:
Private members are rarely used in F#.  Generally in F# this would be a "static let" function (before the public members ) or a helper in an auxiliary module (before the whole class).